### PR TITLE
Allow creating concurrent CAgg refresh policies

### PR DESCRIPTION
**Pull request by kpan2034**
Historically, we only allow creating a single refresh policy on each continuous aggregate.

This PR removes this restriction. We now support creating concurrent CAgg refresh policies, with the restriction that there is no overlap between their refresh intervals.

Along with https://github.com/timescale/timescaledb/pull/8187, this allows refresh policies to run concurrently.

### CODE REVIEW

File: tsl/src/bgw_policy/continuous_aggregate_api.h [hunk 1]
```diff
@@ -6,10 +6,18 @@
 #pragma once
 
 #include <postgres.h>
+#include "bgw_policy/job.h"
 #include "dimension.h"
 #include <continuous_aggs/materialize.h>
 #include <utils/jsonb.h>
 
+typedef enum CaggOverlapResult
```
Comment (fabriziomello) [inline tsl/src/bgw_policy/continuous_aggregate_api.h:14]:
```suggestion
typedef enum CAggOverlapResult
```

File: tsl/src/bgw_policy/continuous_aggregate_api.h [hunk 2]
```diff
@@ -6,10 +6,18 @@
...
+{
+	OVERLAP_NONE,	   /* no overlap */
+	OVERLAP_EXACT,	   /* exact match */
+	OVERLAP_NOT_EXACT, /* overlap but not exact */
+} CaggOverlapResult;
```
Comment (fabriziomello) [inline tsl/src/bgw_policy/continuous_aggregate_api.h]:
```suggestion
typedef enum PolicyRefreshOffsetOverlapResult
{
	POLICY_REFRESH_OFFSET_OVERLAP,		 /* overlap but not equal */
	POLICY_REFRESH_OFFSET_OVERLAP_EQUAL, /* exact match */
	POLICY_REFRESH_OFFSET_OVERLAP_NONE,	 /* don't overlap */
} PolicyRefreshOffsetOverlapResult;
```

File: tsl/src/bgw_policy/continuous_aggregate_api.c [hunk 3]
```diff
@@ -619,6 +727,33 @@ policy_refresh_cagg_add_internal(Oid cagg_oid, Oid start_offset_type, NullableDa
 	JsonbValue *result = pushJsonbValue(&parse_state, WJB_END_OBJECT, NULL);
 	Jsonb *config = JsonbValueToJsonb(result);
 
+	CaggOverlapResult res;
+	res = policy_refresh_cagg_check_for_overlaps(cagg, config, 0);
```
Comment (fabriziomello) [inline tsl/src/bgw_policy/continuous_aggregate_api.c]:
```suggestion
	CAggOverlapResult overlap_result = policy_refresh_cagg_check_for_overlaps(cagg, config, 0);
```

File: tsl/src/bgw_policy/continuous_aggregate_api.c [hunk 4]
```diff
@@ -619,6 +727,33 @@ policy_refresh_cagg_add_internal(Oid cagg_oid, Oid start_offset_type, NullableDa
...
+			{
+				ereport(NOTICE,
+						(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+						 errmsg("continuous aggregate policy already exists for \"%s\", "
+								"skipping",
+								get_rel_name(cagg->relid))));
+				PG_RETURN_INT32(-1);
+			}
+			/* else fallthrough and return error */
+			TS_FALLTHROUGH;
```
Comment (fabriziomello) [inline tsl/src/bgw_policy/continuous_aggregate_api.c]:
```suggestion
			ereport(ERROR,
					(errcode(ERRCODE_DUPLICATE_OBJECT),
					 errmsg("continuous aggregate policy already exists for \"%s\"",
							get_rel_name(cagg_oid)),
					 errdetail("Only one continuous aggregate policy can be created per continuous "
							   "aggregate and a policy with job id %d already exists for \"%s\".",
							   ((BgwJob *) linitial(jobs))->fd.id,
							   get_rel_name(cagg_oid))));
```
I think here we should error out as before

Comment (kpan2034) [reply tsl/src/bgw_policy/continuous_aggregate_api.c]: We still throw an error when there is overlap. We only return a notice when `if_not_exists=true`

File: src/time_utils.c [hunk 5]
```diff
@@ -566,7 +566,7 @@ ts_get_mock_time_or_current_time(void)
 								  Int32GetDatum(-1));
 		return res;
 	}
-	res = TimestampTzGetDatum(GetCurrentTimestamp());
+	res = TimestampTzGetDatum(GetCurrentTransactionStartTimestamp());
```
Comment (fabriziomello) [inline src/time_utils.c:569]: When doing the final rebase let's keep this change into a separate commit similar to d81d1ae8 (in case we need revert it for some reason in the future).

File: tsl/src/bgw_policy/continuous_aggregate_api.c [hunk 6]
```diff
@@ -475,6 +475,165 @@ parse_cagg_policy_config(const ContinuousAgg *cagg, Oid start_offset_type,
...
+
+	Hypertable *mat_ht = ts_hypertable_get_by_id(cagg->data.mat_hypertable_id);
+	const Dimension *dim = get_open_dimension_for_hypertable(mat_ht, true);
+
+	List *jobs = ts_bgw_job_find_by_proc_and_hypertable_id(POLICY_REFRESH_CAGG_PROC_NAME,
+														   FUNCTIONS_SCHEMA_NAME,
+														   cagg->data.mat_hypertable_id);
+
+	if (jobs == NIL)
+		return true;
```
Comment (fabriziomello) [inline tsl/src/bgw_policy/continuous_aggregate_api.c]: I don't think this extra check is necessary since the foreach above will not iterate and then we'll return true anyway

File: tsl/src/bgw_policy/job_api.c [hunk 7]
```diff
@@ -399,6 +400,34 @@ job_alter(PG_FUNCTION_ARGS)
 				 NameStr(job->fd.check_schema),
 				 NameStr(job->fd.check_name));
 
+	// if cagg refresh policy job is being altered, then always check for overlap
+	// there is probably a better place to do this but we choose to do this here
+	// since we need access to the job_id, which we don't have inside the policy_check function
+	// called above
+
```
Comment (fabriziomello) [inline tsl/src/bgw_policy/job_api.c]:
```suggestion
```

File: tsl/src/bgw_policy/continuous_aggregate_api.c (hunk 4 above)
Comment (mkindahl) [reply tsl/src/bgw_policy/continuous_aggregate_api.c]: Yes, but the error message is not very informative for this case. Better to say that it is a duplicate than to say that it is an overlap.

File: tsl/src/bgw_policy/continuous_aggregate_api.c [hunk 8]
```diff
@@ -619,6 +724,32 @@ policy_refresh_cagg_add_internal(Oid cagg_oid, Oid start_offset_type, NullableDa
...
+		case CAGG_OVERLAP_NONE:
+			break;
+		case CAGG_OVERLAP_EXACT:
+			if (if_not_exists)
+			{
+				ereport(NOTICE,
+						(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+						 errmsg("continuous aggregate policy already exists for \"%s\", "
+								"skipping",
+								get_rel_name(cagg->relid))));
```
Comment (mkindahl) [inline tsl/src/bgw_policy/continuous_aggregate_api.c]: Suggest to add some more information to explain how you evaluate if there is already a continuous aggregate.
```suggestion
				ereport(NOTICE,
						(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
						 errmsg("an identical continuous aggregate policy already exists for \"%s\", "
								"skipping",
								get_rel_name(cagg->relid)),
						errdetail("A policy with the same start and end offset already exists for continuous aggregate \"%s\"", get_rel_name(cagg->relid))));
```

File: tsl/src/bgw_policy/continuous_aggregate_api.c (hunk 4 above)
Comment (mkindahl) [reply tsl/src/bgw_policy/continuous_aggregate_api.c]: Suggest to add to the errdetail that it has the same start and end offset so that it is clear how the decision if policies are "identical" is made. See my suggestion above.

File: tsl/src/bgw_policy/continuous_aggregate_api.c (hunk 8 above)
Comment (kpan2034) [reply tsl/src/bgw_policy/continuous_aggregate_api.c]: Changed. I also changed the error message when `if_force_exists=false` but an identical policy exists, to improve clarity.

**Comment by kpan2034** [comment]
> Looks good. Some suggestions for improving the error message, and it is not clear if there are tests for these cases:
> 
> * Add a policy with identical start and end offset.
> * Alter a policy so that it overlaps with an existing policy
> * Alter a policy so that it has identical start and end offset with an existing policy.

The first case is already covered. I added tests for the other two cases.

File: tsl/src/bgw_policy/continuous_aggregate_api.c [hunk 9]
```diff
@@ -5,13 +5,15 @@
  */
 
 #include <postgres.h>
+
+#include "annotations.h"
```
Comment (fabriziomello) [inline tsl/src/bgw_policy/continuous_aggregate_api.c]: do we need this header? and if we need would be good to move below where we include tsdb headers, the current convention is:
1. postgres.h
2. other postgres headers
3. tsdb headers

File: tsl/test/expected/cagg_policy.out [hunk 10]
```diff
@@ -232,19 +232,13 @@ DETAIL:  The start and end offsets must cover at least two buckets in the valid
 SELECT add_continuous_aggregate_policy('mat_m1', 20, 10, '1h'::interval) as job_id \gset
 --adding again should warn/error
 SELECT add_continuous_aggregate_policy('mat_m1', 20, 10, '1h'::interval, if_not_exists=>false);
-ERROR:  continuous aggregate policy already exists for "mat_m1"
-DETAIL:  Only one continuous aggregate policy can be created per continuous aggregate and a policy with job id 1009 already exists for "mat_m1".
+ERROR:  an identical continuous aggregate policy already exists for "mat_m1", skipping
```
Comment (fabriziomello) [inline tsl/test/expected/cagg_policy.out]:
```suggestion
ERROR:  continuous aggregate refresh policy already exists for "mat_m1", skipping
```

File: tsl/test/expected/cagg_policy.out [hunk 11]
```diff
@@ -232,19 +232,13 @@ DETAIL:  The start and end offsets must cover at least two buckets in the valid
...
+DETAIL:  A policy with the same start and end offset already exists for continuous aggregate "mat_m1"
```
Comment (fabriziomello) [inline tsl/test/expected/cagg_policy.out]:
```suggestion
DETAIL:  A refresh policy with the same start and end offset already exists for continuous aggregate "mat_m1"
```

File: tsl/src/bgw_policy/continuous_aggregate_api.c [hunk 12]
```diff
@@ -484,6 +484,162 @@ parse_cagg_policy_config(const ContinuousAgg *cagg, Oid start_offset_type,
...
+		else if (range_overlaps_internal(typcache, range_job, range))
+		{
+			overlaps = true;
+		}
+	}
+
+	if (overlaps)
+	{
+		return CAGG_OVERLAP_NOT_EXACT;
+	}
```
Comment (fabriziomello) [inline tsl/src/bgw_policy/continuous_aggregate_api.c]:
```suggestion
		else if (range_overlaps_internal(typcache, range_job, range))
		{
			return CAGG_OVERLAP_NOT_EXACT;
		}
	}
```

File: tsl/src/bgw_policy/continuous_aggregate_api.c [hunk 13]
```diff
@@ -628,6 +733,46 @@ policy_refresh_cagg_add_internal(Oid cagg_oid, Oid start_offset_type, NullableDa
...
+			{
+				ereport(ERROR,
+						(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+						 errmsg("an identical continuous aggregate policy already exists for "
+								"\"%s\", ",
+								get_rel_name(cagg->relid)),
+						 errdetail("A policy with the same start and end offset already exists for "
+								   "continuous aggregate \"%s\"",
+								   get_rel_name(cagg->relid))));
+			}
```
Comment (fabriziomello) [inline tsl/src/bgw_policy/continuous_aggregate_api.c]:
```suggestion

				ereport(ERROR,
						(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
						 errmsg("an identical continuous aggregate policy already exists for "
								"\"%s\", ",
								get_rel_name(cagg->relid)),
						 errdetail("A policy with the same start and end offset already exists for "
								   "continuous aggregate \"%s\"",
								   get_rel_name(cagg->relid))));
```
Don't need the `else` branch.

File: tsl/src/bgw_policy/continuous_aggregate_api.c (hunk 12 above)
Comment (kpan2034) [reply tsl/src/bgw_policy/continuous_aggregate_api.c]: We can't do it like this since we first need to check _all_ existing jobs for an exact match. We prefer returning an exact match over an overlap, and the list of jobs we iterate over isn't guaranteed to be sorted by start/end offset.

File: tsl/src/bgw_policy/continuous_aggregate_api.c (hunk 9 above)
Comment (kpan2034) [reply tsl/src/bgw_policy/continuous_aggregate_api.c]: Good catch, was a leftover from a previous version of this PR where we used `TS_FALLTHROUGH`. Removed.

File: tsl/test/expected/cagg_policy.out (hunk 10 above)
Comment (kpan2034) [reply tsl/test/expected/cagg_policy.out]: Changed wording

File: tsl/src/bgw_policy/continuous_aggregate_api.c (hunk 12 above)
Comment (fabriziomello) [reply tsl/src/bgw_policy/continuous_aggregate_api.c]: You're correct.

File: tsl/src/bgw_policy/continuous_aggregate_api.h [hunk 14]
```diff
@@ -6,10 +6,18 @@
...
 #include <continuous_aggs/materialize.h>
 #include <utils/jsonb.h>
 
+typedef enum CAggOverlapResult
+{
+	CAGG_OVERLAP_NONE,		/* no overlap */
+	CAGG_OVERLAP_EXACT,		/* exact match */
+	CAGG_OVERLAP_NOT_EXACT, /* overlap but not exact */
+} CAggOverlapResult;
+
```
Comment (fabriziomello) [inline tsl/src/bgw_policy/continuous_aggregate_api.h]: https://github.com/timescale/timescaledb/pull/8274#discussion_r2172967123

Comment (kpan2034) [reply tsl/src/bgw_policy/continuous_aggregate_api.h]: Updated

File: tsl/src/bgw_policy/continuous_aggregate_api.c [hunk 15]
```diff
@@ -484,6 +482,162 @@ parse_cagg_policy_config(const ContinuousAgg *cagg, Oid start_offset_type,
...
+		else if (range_overlaps_internal(typcache, range_job, range))
+		{
+			overlaps = true;
+		}
+	}
+
+	if (overlaps)
+	{
+		return POLICY_REFRESH_OFFSET_OVERLAP_NONE;
+	}
```
Comment (mkindahl) [inline tsl/src/bgw_policy/continuous_aggregate_api.c]: Is this correct? Shouldn't it be the opposite?

```suggestion
	if (!overlaps)
	{
		return POLICY_REFRESH_OFFSET_OVERLAP_NONE;
	}
```

File: tsl/src/bgw_policy/continuous_aggregate_api.c [hunk 16]
```diff
@@ -484,6 +482,162 @@ parse_cagg_policy_config(const ContinuousAgg *cagg, Oid start_offset_type,
...
+		if (start_offset == start_offset_job && end_offset == end_offset_job)
+		{
+			/* If all arguments are the same, do nothing */
+			return POLICY_REFRESH_OFFSET_OVERLAP_EQUAL;
+		}
+		else if (range_overlaps_internal(typcache, range_job, range))
+		{
+			overlaps = true;
+		}
+	}
```
Comment (mkindahl) [inline tsl/src/bgw_policy/continuous_aggregate_api.c]: Suggest to add a comment here that you need to check all policies and see if you have an exact match.

File: tsl/src/bgw_policy/continuous_aggregate_api.h [hunk 17]
```diff
@@ -6,10 +6,18 @@
...
 #include "dimension.h"
 #include <continuous_aggs/materialize.h>
 #include <utils/jsonb.h>
 
+typedef enum PolicyRefreshOffsetOverlapResult
+{
+	POLICY_REFRESH_OFFSET_OVERLAP,		 /* no overlap */
+	POLICY_REFRESH_OFFSET_OVERLAP_EQUAL, /* exact match */
+	POLICY_REFRESH_OFFSET_OVERLAP_NONE,	 /* overlap but not exact */
+} PolicyRefreshOffsetOverlapResult;
```
Comment (mkindahl) [inline tsl/src/bgw_policy/continuous_aggregate_api.h]: This is very confusing. I expected the `POLICY_REFRESH_OFFSET_OVERLAP_NONE` to be "no overlap" and `POLICY_REFRESH_OFFSET_OVERLAP` to mean there is an overlap but not exact.

File: tsl/src/bgw_policy/continuous_aggregate_api.c (hunk 15 above)
Comment (mkindahl) [reply tsl/src/bgw_policy/continuous_aggregate_api.c]: I see that the enum is using a different naming than I expected.

File: tsl/src/bgw_policy/continuous_aggregate_api.c [hunk 18]
```diff
@@ -628,6 +731,45 @@ policy_refresh_cagg_add_internal(Oid cagg_oid, Oid start_offset_type, NullableDa
...
+								   "continuous aggregate \"%s\"",
+								   get_rel_name(cagg->relid))));
+				PG_RETURN_INT32(-1);
+			}
+			ereport(ERROR,
+					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+
+					 errmsg("continuous aggregate refresh policy already exists for "
+							"\"%s\", ",
+							get_rel_name(cagg->relid)),
```
Comment (mkindahl) [inline tsl/src/bgw_policy/continuous_aggregate_api.c]: Suggest you use the same wording for both the notice and the error, with the exception of the "skipping" last. Also, a typo.
```suggestion
					 errmsg("an identical continuous aggregate refresh policy already exists for "
							"\"%s\"",
							get_rel_name(cagg->relid)),
```

File: tsl/test/expected/cagg_policy_concurrent.out [hunk 19]
```diff
@@ -0,0 +1,1368 @@
...
+
+\set ON_ERROR_STOP 0
+/* Alter end offset to overlap with another job*/
+SELECT jsonb_set(:'config', '{end_offset}', '1000') AS config \gset
+SELECT * FROM alter_job(:job_id, config := :'config');
+ERROR:  refresh interval overlaps with an existing continuous aggregate policy on "mat_m1"
+/* Alter end offset to be null */
+SELECT jsonb_set(:'config', '{end_offset}', 'null') AS config \gset
+SELECT * FROM alter_job(:job_id, config := :'config');
+ERROR:  refresh interval overlaps with an existing continuous aggregate policy on "mat_m1"
```
Comment (mkindahl) [inline tsl/test/expected/cagg_policy_concurrent.out:345]: Also test the case that you alter the "last" job so that the end range is null but does not overlap anything else.

**Review by mkindahl — COMMENTED** [review]
One request to change the name of the enums since that will be very confusing as well as one additional test. The rest are nits.

File: tsl/src/bgw_policy/continuous_aggregate_api.c [hunk 20]
```diff
@@ -628,6 +731,45 @@ policy_refresh_cagg_add_internal(Oid cagg_oid, Oid start_offset_type, NullableDa
...
+			if (if_not_exists)
+			{
+				ereport(NOTICE,
+						(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+						 errmsg("an identical continuous aggregate policy already exists for "
+								"\"%s\", "
+								"skipping",
+								get_rel_name(cagg->relid)),
+						 errdetail("A policy with the same start and end offset already exists for "
+								   "continuous aggregate \"%s\"",
```
Comment (erimatnor) [inline tsl/src/bgw_policy/continuous_aggregate_api.c]:
```suggestion
						 errdetail("A policy with the same start and end offset already exists for "
								   "continuous aggregate \"%s\".",
```

File: tsl/src/bgw_policy/continuous_aggregate_api.c [hunk 21]
```diff
@@ -628,6 +731,45 @@ policy_refresh_cagg_add_internal(Oid cagg_oid, Oid start_offset_type, NullableDa
...
+			}
+			ereport(ERROR,
+					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+
+					 errmsg("continuous aggregate refresh policy already exists for "
+							"\"%s\", ",
+							get_rel_name(cagg->relid)),
+					 errdetail("A refresh policy with the same start and end offset already exists "
+							   "for "
+							   "continuous aggregate \"%s\"",
```
Comment (erimatnor) [inline tsl/src/bgw_policy/continuous_aggregate_api.c]:
```suggestion
					 errdetail("A refresh policy with the same start and end offset already exists "
							   "for "
							   "continuous aggregate \"%s\".",
```

File: tsl/src/bgw_policy/job_api.c [hunk 22]
```diff
@@ -399,6 +400,45 @@ job_alter(PG_FUNCTION_ARGS)
...
+				ereport(ERROR,
+						(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+
+						 errmsg("continuous aggregate refresh policy already exists for "
+								"\"%s\", ",
+								get_rel_name(cagg->relid)),
+						 errdetail("A refresh policy with the same start and end offset already "
+								   "exists for "
+								   "continuous aggregate \"%s\"",
+								   get_rel_name(cagg->relid))));
```
Comment (erimatnor) [inline tsl/src/bgw_policy/job_api.c]:
```suggestion
						 errdetail("A refresh policy with the same start and end offset already "
								   "exists for "
								   "continuous aggregate \"%s\".",
								   get_rel_name(cagg->relid))));
```

File: tsl/src/bgw_policy/continuous_aggregate_api.h (hunk 17 above)
Comment (fabriziomello) [reply tsl/src/bgw_policy/continuous_aggregate_api.h]: You're correct Mats, the renaming lead to reordering the items.

File: tsl/src/bgw_policy/continuous_aggregate_api.c (hunk 15 above)
Comment (kpan2034) [reply tsl/src/bgw_policy/continuous_aggregate_api.c]: Fixed enum naming

File: tsl/src/bgw_policy/continuous_aggregate_api.h (hunk 17 above)
Comment (kpan2034) [reply tsl/src/bgw_policy/continuous_aggregate_api.h]: Fixed the naming. It should behave as you expected.

File: tsl/src/bgw_policy/continuous_aggregate_api.c (hunk 18 above)
Comment (kpan2034) [reply tsl/src/bgw_policy/continuous_aggregate_api.c]: Changed the message for the NOTICE to be the same as for the ERROR.

File: tsl/test/expected/cagg_policy_concurrent.out (hunk 19 above)
Comment (kpan2034) [reply tsl/test/expected/cagg_policy_concurrent.out:345]: Added this test case.

File: tsl/src/bgw_policy/continuous_aggregate_api.c [hunk 23]
```diff
@@ -484,6 +482,165 @@ parse_cagg_policy_config(const ContinuousAgg *cagg, Oid start_offset_type,
...
+}
+
+/* Ensures the refresh range of the new policy doesn't overlap with an existing one*/
+PolicyRefreshOffsetOverlapResult
+policy_refresh_cagg_check_for_overlaps(ContinuousAgg *cagg, Jsonb *policy_config,
+									   int32 existing_job_id)
+{
+	List *jobs = ts_bgw_job_find_by_proc_and_hypertable_id(POLICY_REFRESH_CAGG_PROC_NAME,
+														   FUNCTIONS_SCHEMA_NAME,
+														   cagg->data.mat_hypertable_id);
```
Comment (fabriziomello) [inline tsl/src/bgw_policy/continuous_aggregate_api.c:530]:
```suggestion
	List *jobs = ts_bgw_job_find_by_proc_and_hypertable_id(POLICY_REFRESH_CAGG_PROC_NAME,
														   FUNCTIONS_SCHEMA_NAME,
														   cagg->data.mat_hypertable_id);
	PolicyRefreshOffsetOverlapResult overlap_result = POLICY_REFRESH_OFFSET_OVERLAP_NONE;
```

File: tsl/src/bgw_policy/continuous_aggregate_api.c [hunk 24]
```diff
@@ -484,6 +482,165 @@ parse_cagg_policy_config(const ContinuousAgg *cagg, Oid start_offset_type,
...
+
+	if (jobs == NIL)
+		return POLICY_REFRESH_OFFSET_OVERLAP_NONE;
```
Comment (fabriziomello) [inline tsl/src/bgw_policy/continuous_aggregate_api.c]:
```suggestion
		return overlap_result;
```

File: tsl/src/bgw_policy/continuous_aggregate_api.c [hunk 25]
```diff
@@ -484,6 +482,165 @@ parse_cagg_policy_config(const ContinuousAgg *cagg, Oid start_offset_type,
...
+	if (typcache == NULL || typcache->rngelemtype == NULL)
+		elog(ERROR, "cache lookup failed");
+
+#if PG16_LT
+	RangeType *range = make_range(typcache, &lower, &upper, false);
+#else
+	RangeType *range = make_range(typcache, &lower, &upper, false, NULL);
+#endif
+
+	bool overlaps = false;
```
Comment (fabriziomello) [inline tsl/src/bgw_policy/continuous_aggregate_api.c]:
```suggestion
```

File: tsl/src/bgw_policy/continuous_aggregate_api.c [hunk 26]
```diff
@@ -484,6 +482,165 @@ parse_cagg_policy_config(const ContinuousAgg *cagg, Oid start_offset_type,
...
+		{
+			/* If all arguments are the same, do nothing */
+			return POLICY_REFRESH_OFFSET_OVERLAP_EQUAL;
+		}
+		/* We need to first check all other jobs to see if there is an exact match, since we prefer
+		 * returning an exact match over an overlap, and the list of jobs isn't guaranteed to be
+		 * sorted by start/end offset. */
+		else if (range_overlaps_internal(typcache, range_job, range))
+		{
+			overlaps = true;
```
Comment (fabriziomello) [inline tsl/src/bgw_policy/continuous_aggregate_api.c]:
```suggestion
			overlap_result = POLICY_REFRESH_OFFSET_OVERLAP;
```

File: tsl/src/bgw_policy/continuous_aggregate_api.c [hunk 27]
```diff
@@ -484,6 +482,165 @@ parse_cagg_policy_config(const ContinuousAgg *cagg, Oid start_offset_type,
...
+		}
+	}
+
+	if (overlaps)
+	{
+		return POLICY_REFRESH_OFFSET_OVERLAP;
+	}
+
+	return POLICY_REFRESH_OFFSET_OVERLAP_NONE;
```
Comment (fabriziomello) [inline tsl/src/bgw_policy/continuous_aggregate_api.c]:
```suggestion
	return overlap_result;
```
